wrapanapi/containers/providers/rhopenshift.py

```diff
@@ -76,6 +76,34 @@ def retrieve_cfme_appliance_version(template_name):
 """
 
 
+def reconnect(decorator):
+    def decorate(cls):
+        for attr in cls.__dict__:
+            if callable(getattr(cls, attr)) and not attr.startswith('_'):
+                setattr(cls, attr, decorator(getattr(cls, attr)))
+        return cls
+    return decorate
+
+
+def unauthenticated_error_handler(method):
+    """Fixes issue with 401 error by restoring connection.
+        Sometimes connection to openshift api endpoint gets expired and openshift returns 401.
+        As a result tasks in some applications like sprout fail.
+    """
+    def wrap(*args, **kwargs):
+        attempts = 3
+        for attempt in range(1, attempts + 1):
+            try:
+                return method(*args, **kwargs)
+            except ApiException, e:
+                if e.reason == 'Unauthorized' and attempt < attempts:
+                    args[0].connect()
+                else:
+                    raise e
+    return wrap
+
+
+@reconnect(unauthenticated_error_handler)
```

Comment:
form maintenance it might be better to actually apply the decorator on each method

Reply:
It isn't difficult to me to do taht. however, there are very many methods and their amount will grow up. Do you think it is necessary ?